sys/include/wtimer.h

```diff
@@ -0,0 +1,330 @@
+void wtimer_set(wtimer_t *timer, uint32_t offset);
+
+/**
+ * @brief wtimer backoff value
+ *
+ * All timers that are less than WTIMER_BACKOFF microseconds in the future will
+ * just spin.
+ *
+ * This is supposed to be defined per-device in e.g., periph_conf.h.
+ */
+#ifndef WTIMER_BACKOFF
+#define WTIMER_BACKOFF 30
+#endif
+
+/**
+ * @brief wtimer overhead value
+ *
+ * This value specifies the time a timer will be late if uncorrected.
+ *
+ * E.g., with WTIMER_OVERHEAD == 0
+ * now=wtimer_now();
+ * wtimer_set(&timer, X);
+ * overhead=X-now
+ *
+ * wtimer automatically substracts WTIMER_OVERHEAD from a timer's target time.
+ *
+ * This is supposed to be defined per-device in e.g., periph_conf.h.
+ */
+#ifndef WTIMER_OVERHEAD
+#define WTIMER_OVERHEAD 20
+#endif
+
+#ifndef WTIMER_ISR_BACKOFF
+#define WTIMER_ISR_BACKOFF 20
+#endif
+
+#ifndef WTIMER_MASK
+/**
+ * @brief wtimer timer mask
+ *
+ * This value specifies the mask relative to 0xffffffff that the used timer
+ * counts to, e.g., 0xffffffff & ~TIMER_MAXVALUE.
+ *
+ * For a 16bit timer, the mask would be 0xFFFF0000, for a 24bit timer, the mask
+ * would be 0xFF000000. Don't set this for 32bit timers.
+ *
+ * This is supposed to be defined per-device in e.g., periph_conf.h.
+ */
+#define WTIMER_MASK 0
+#endif
+
+#ifndef WTIMER_USLEEP_UNTIL_OVERHEAD
+/**
+ * @brief wtimer_usleep_until overhead value
+ *
+ * This value specifies the time a wtimer_usleep_until will be late
+ * if uncorrected.
+ *
+ * This is supposed to be defined per-device in e.g., periph_conf.h.
+ */
+#define WTIMER_USLEEP_UNTIL_OVERHEAD 10
+#endif
+
+/**
+ * @{
+ * @brief wtimer internal stuff
+ */
+#if WTIMER_MASK
+extern volatile uint32_t _high_cnt;
+#endif
+
+/**
+ * @brief returns the (masked) low-level timer counter value.
+ */
+static inline uint32_t _wtimer_now(void)
+{
+    return timer_read(WTIMER);
+}
+
+/**
+ * @brief drop bits of a value that don't fit into the low-level timer.
+ */
+static inline uint32_t _mask(uint32_t val)
+{
+    return val & ~WTIMER_MASK;
+}
+
+int _wtimer_set_absolute(wtimer_t *timer, uint32_t target);
+void _wtimer_set64(wtimer_t *timer, uint32_t offset, uint32_t long_offset);
+void _wtimer_sleep(uint32_t offset, uint32_t long_offset);
+/** @} */
+
+static inline uint32_t wtimer_now(void)
+{
+#if WTIMER_MASK
+    return _wtimer_now() | _high_cnt;
+#else
+    return _wtimer_now();
+#endif
+}
+
+static inline void wtimer_spin_until(uint32_t value) {
+    while (_wtimer_now() > value);
+    while (_wtimer_now() < value);
+}
+
+static inline void wtimer_spin(uint32_t offset) {
+    offset = _mask(offset + _wtimer_now());
+    wtimer_spin_until(offset);
+}
+
+static inline void wtimer_usleep(uint32_t offset)
+{
+    _wtimer_sleep(offset, 0);
+}
+
+static inline void wtimer_usleep64(uint64_t microseconds)
+{
+    _wtimer_sleep((uint32_t) microseconds, (uint32_t) (microseconds >> 32));
+}
+
+static inline void wtimer_sleep(uint32_t seconds)
+{
+    wtimer_usleep64((uint64_t)seconds*1000000);
```

Comment:
I think `1000000` must at least be written `1000000ul` on avr in order to make it a 32 bit value, unless the compiler is smart enough to notice the error.